source/extensions/config_subscription/grpc/new_grpc_mux_impl.cc

```diff
@@ -246,6 +252,41 @@ GrpcMuxWatchPtr NewGrpcMuxImpl::addWatch(const std::string& type_url,
   return std::make_unique<WatchImpl>(type_url, watch, *this, options);
 }
 
+absl::Status
+NewGrpcMuxImpl::updateMuxSource(Grpc::RawAsyncClientPtr primary_async_client,
+                                Grpc::RawAsyncClientPtr failover_async_client,
+                                CustomConfigValidatorsPtr custom_config_validators,
+                                Stats::Scope& scope, BackOffStrategyPtr backoff_strategy,
+                                const envoy::config::core::v3::ApiConfigSource& ads_config_source) {
+  // Process the rate limit settings.
+  absl::StatusOr<RateLimitSettings> rate_limit_settings_or_error =
+      Utility::parseRateLimitSettings(ads_config_source);
+  RETURN_IF_NOT_OK_REF(rate_limit_settings_or_error.status());
+
+  const Protobuf::MethodDescriptor& service_method =
+      *Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "envoy.service.discovery.v3.AggregatedDiscoveryService.DeltaAggregatedResources");
+
+  // Disconnect from current xDS servers.
+  ENVOY_LOG_MISC(info, "Replacing the xDS gRPC mux source");
+  grpc_stream_->closeStream();
+  grpc_stream_ = createGrpcStreamObject(std::move(primary_async_client),
+                                        std::move(failover_async_client), service_method, scope,
+                                        std::move(backoff_strategy), *rate_limit_settings_or_error);
+
+  // Update the config validators.
+  config_validators_ = std::move(custom_config_validators);
+  // Update the watch map's config validators.
+  for (auto& [type_url, subscription] : subscriptions_) {
+    subscription->watch_map_.setConfigValidators(config_validators_.get());
+  }
+
+  // Start the subscriptions over the grpc_stream.
+  grpc_stream_->establishNewStream();
+
+  return absl::OkStatus();
```

Comment:
this method does not guarantee that stream establishment was successful? E.g. it attempts to establish new stream and returns `absl::OkStatus()` independently if attempt was successful or not.

Reply:
No, the method does not guarantee that the stream establishment was successful.
The model of the gRPC-Muxes is as follows:
The gRPC-Mux receives a grpc-async-client, creates a grpc-stream (or grpc-failover) object and delegates the connection (stream establishment) and reconnection to that object.